### PR TITLE
[Example]RGCN new-sampler example using homogeneous run for heterogeneous graph.

### PR DESCRIPTION
**Pull request by classicsong**
## Description
This PR shows how to using new sampler API to implement mini-batch based RGCN under homogeneous graph. Following features are provided:
 - neighbor sample based mini-batch training for both node classification and link prediction
 - multi-gpu training for node classification and link prediction
 - add new link prediction example dataset [DRKG](https://github.com/gnn4dr/DRKG)

Node classification speed (time per epoch * num of epochs)
   | Full graph | MinibatchHomogenous 
-- | -- | -- 
Aibf | 5.35ms * 50 | 16.9ms  * 50
Mutag | 6.06ms * 50 | 46.3ms * 30
Bgs | 2.54ms * 50 | 8.52ms * 60
Am | 64.8s * 50 | 46.9ms * 30

Link prediction speed (time per epoch * num of epochs):
   | Full graph | MinibatchHomogenous 
-- | -- | -- 
wn18-train |  0.1s * 4000 | 19.56s  * 30
wn18-test | 194.13s | 88.9s
wn18-test memory cost | 60 GB | Fit in GPU

testbed: p3.2xlarge
Node classification multigpu-training (time per epoch):
   | 1GPU | 4 GPU
-- | -- | -- 
AM-train | 46.9ms | 12.2ms

link prediction multigpu-training (time per epoch):
   | 1GPU | 4 GPU
-- | -- | -- 
wn18-train | 19.56s | 5.09s
wn18-test | 88.9s | 33.3s

testbed: p3.8xlarge

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/rgcn/dataset.py [hunk 1]
```diff
@@ -0,0 +1,535 @@
...
+import pandas as pd
+from collections import Counter
+import re
+
+from dgl.data.utils import download, extract_archive, get_download_dir, _get_dgl_url
+
+np.random.seed(123)
+_downlaod_prefix = _get_dgl_url('dataset/')
+
+class RGCNEntityDataset(object):
```
Comment (zheng-da) [inline examples/pytorch/rgcn/dataset.py:22]: why do we need another data loader? i thought we loaded these graphs in both DGLGraph and DGLHeteroGraph.

File: examples/pytorch/rgcn/entity_classify_hetero_mb.py [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+        norms = []
+        ntypes = []
+        seeds = th.tensor(seeds).long()
+        cur = self.target_idx[seeds]
+        for fanout in self.fanouts:
+            if fanout is None or fanout == -1:
+                frontier = dgl.in_subgraph(self.g, cur)
+            else:
+                frontier = dgl.sampling.sample_neighbors(self.g, cur, fanout)
+            etypes = self.g.edata[dgl.ETYPE][frontier.edata[dgl.EID]]
```
Comment (zheng-da) [inline examples/pytorch/rgcn/entity_classify_hetero_mb.py:154]: do we need `dgl.ETYPE`? i suppose `g` has only one relation.

File: examples/pytorch/rgcn/dataset.py (hunk 1 above)
Comment (classicsong) [reply examples/pytorch/rgcn/dataset.py:22]: This allows loading data in DGLHeteroGraph and use to_homo to generate homogeneous graph.

File: examples/pytorch/rgcn/utils.py [hunk 3]
```diff
@@ -334,3 +343,326 @@ def calc_mrr(embedding, w, train_triplets, valid_triplets, test_triplets, hits=[
...
+        self.valid = (valid_head, valid_tail, valid_head_type, valid_rel_type, valid_tail_type)
+        self.test = (test_head, test_tail, test_head_type, test_rel_type, test_tail_type)
+        self.rel_map = relation_dict
+        self.entity_map = entity_dict
+        self.entity_type_dict = entity_type_dict
+        self.num_rels = len(list(graph_relations))
+        self.num_nodes = len(entity_dict)
+        self.node_types = np.asarray(node_types)
+
+def build_multi_ntype_heterograph_in_homogeneous_from_triplets(num_nodes, num_rels, node_types, edge_lists, reverse=True):
```
Comment (jermainewang) [inline examples/pytorch/rgcn/utils.py:464]: Is this the same as `dgl.to_homo`?

File: examples/pytorch/rgcn/utils.py [hunk 4]
```diff
@@ -334,3 +343,326 @@ def calc_mrr(embedding, w, train_triplets, valid_triplets, test_triplets, hits=[
...
+    return g
+
+#######################################################################
+#
+# Build Heterograph in Homogeneous Graph for Link Prediction
+#
+#######################################################################
+def build_heterograph_in_homogeneous_from_triplets(num_nodes, num_rels, edge_lists, reverse=True):
+    """ Create a DGL homogeneous graph with heterograph info stored as node or edge features.
+    """
```
Comment (jermainewang) [inline examples/pytorch/rgcn/utils.py:560]: What's the difference between this function and the one above?

File: examples/pytorch/rgcn/README.md [hunk 5]
```diff
@@ -37,6 +37,33 @@ AM: accuracy 87.37% (DGL), 89.29% (paper)
 python3 entity_classify.py -d am --n-bases=40 --n-hidden=10 --l2norm=5e-4 --testing
 ```
 
+#### Entity Classification with minibatch
+AIFB: accuracy avg(5 runs) 94.44%, best 94.44% (DGL)
+```
+python3 entity_classify_hetero_mb.py -d aifb --testing --gpu 0 --fanout=20 --batch-size 128
```
Comment (jermainewang) [inline examples/pytorch/rgcn/README.md:43]: `entity_classify_hetero_mb` -> `entity_classify_mb`

File: examples/pytorch/rgcn/utils.py (hunk 3 above)
Comment (classicsong) [reply examples/pytorch/rgcn/utils.py:464]: to_homo may not work for link prediction. There will be three graphs train_g, valid_g, test_g
Or I did not find a easy way to use to_homo.

File: examples/pytorch/rgcn/utils.py (hunk 4 above)
Comment (classicsong) [reply examples/pytorch/rgcn/utils.py:560]: One for node with type and another node has no type.
